fix #13837 the Invalid log : worker worker have not received the hear…

lxorc opened the pull request:
fix #13837 the Invalid log : worker worker have not received the hearbet

lxorc committed:
fix #13837 the Invalid log : worker worker have not received the heartbeat

ruanwenjun commented:
We don't need to add this check, this case might be due to we registering an empty node? Could you please describe how to reproduce this.

lxorc commented:
> We don't need to add this check, this case might be due to we registering an empty node? Could you please describe how to reproduce this.

Yes, by default a `worker` node will be loaded in `registryClient.subscribe(REGISTRY_DOLPHINSCHEDULER_WORKERS, new WorkerDataListener()); `

```java
registryClient.subscribe(REGISTRY_DOLPHINSCHEDULER_WORKERS, new WorkerDataListener()); 
String[] parts = path.split("/");  // REGISTRY_DOLPHINSCHEDULER_WORKERS = /nodes/worker
final String workerAddress = parts[parts.length - 1];  // workerAddress  = worker
if (type == Type.UPDATE) // the Type.UPDATE is  Node /nodes/worker update ,   case NODE_UPDATED: type(Type.UPDATE);
syncSingleWorkerNodeInfo(workerAddress, JSONUtils.parseObject(data, WorkerHeartBeat.class));  //  ???
workerNodeInfo.put(workerAddress, info);  // put "worker" key into the workerNodeInfo
```


The effect is that the `addr` of `getHostWeight` will be the `worker` and `heartBeat == null`.

so 

```java
        if (heartBeat == null) {
            logger.warn("worker {} in work group {} have not received the heartbeat", addr, workerGroup);
            return Optional.empty();
        }
```

ruanwenjun commented:
> > We don't need to add this check, this case might be due to we registering an empty node? Could you please describe how to reproduce this.
> 
> Yes, by default a `worker` node will be loaded in `registryClient.subscribe(REGISTRY_DOLPHINSCHEDULER_WORKERS, new WorkerDataListener()); `
> 
> ```java
> registryClient.subscribe(REGISTRY_DOLPHINSCHEDULER_WORKERS, new WorkerDataListener()); 
> String[] parts = path.split("/");  // REGISTRY_DOLPHINSCHEDULER_WORKERS = /nodes/worker
> final String workerAddress = parts[parts.length - 1];  // workerAddress  = worker
> if (type == Type.UPDATE) // the Type.UPDATE is  Node /nodes/worker update ,   case NODE_UPDATED: type(Type.UPDATE);
> syncSingleWorkerNodeInfo(workerAddress, JSONUtils.parseObject(data, WorkerHeartBeat.class));  //  ???
> workerNodeInfo.put(workerAddress, info);  // put "worker" key into the workerNodeInfo
> ```
> 
> The effect is that the `addr` of `getHostWeight` will be the `worker` and `heartBeat == null`.
> 
> so
> 
> ```java
>         if (heartBeat == null) {
>             logger.warn("worker {} in work group {} have not received the heartbeat", addr, workerGroup);
>             return Optional.empty();
>         }
> ```

So we only need to subscribe the event of REGISTRY_DOLPHINSCHEDULER_WORKERS's child path

lxorc commented:
> > > We don't need to add this check, this case might be due to we registering an empty node? Could you please describe how to reproduce this.
> > 
> > 
> > Yes, by default a `worker` node will be loaded in `registryClient.subscribe(REGISTRY_DOLPHINSCHEDULER_WORKERS, new WorkerDataListener()); `
> > ```java
> > registryClient.subscribe(REGISTRY_DOLPHINSCHEDULER_WORKERS, new WorkerDataListener()); 
> > String[] parts = path.split("/");  // REGISTRY_DOLPHINSCHEDULER_WORKERS = /nodes/worker
> > final String workerAddress = parts[parts.length - 1];  // workerAddress  = worker
> > if (type == Type.UPDATE) // the Type.UPDATE is  Node /nodes/worker update ,   case NODE_UPDATED: type(Type.UPDATE);
> > syncSingleWorkerNodeInfo(workerAddress, JSONUtils.parseObject(data, WorkerHeartBeat.class));  //  ???
> > workerNodeInfo.put(workerAddress, info);  // put "worker" key into the workerNodeInfo
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The effect is that the `addr` of `getHostWeight` will be the `worker` and `heartBeat == null`.
> > so
> > ```java
> >         if (heartBeat == null) {
> >             logger.warn("worker {} in work group {} have not received the heartbeat", addr, workerGroup);
> >             return Optional.empty();
> >         }
> > ```
> 
> So we only need to subscribe the event of REGISTRY_DOLPHINSCHEDULER_WORKERS's child path

Please help to review my code to see if there are other problems. If there is a better implementation, please reply to me. And I have fixed it in my production environment and it works fine.

ruanwenjun commented:
> > > > We don't need to add this check, this case might be due to we registering an empty node? Could you please describe how to reproduce this.
> > > 
> > > 
> > > Yes, by default a `worker` node will be loaded in `registryClient.subscribe(REGISTRY_DOLPHINSCHEDULER_WORKERS, new WorkerDataListener()); `
> > > ```java
> > > registryClient.subscribe(REGISTRY_DOLPHINSCHEDULER_WORKERS, new WorkerDataListener()); 
> > > String[] parts = path.split("/");  // REGISTRY_DOLPHINSCHEDULER_WORKERS = /nodes/worker
> > > final String workerAddress = parts[parts.length - 1];  // workerAddress  = worker
> > > if (type == Type.UPDATE) // the Type.UPDATE is  Node /nodes/worker update ,   case NODE_UPDATED: type(Type.UPDATE);
> > > syncSingleWorkerNodeInfo(workerAddress, JSONUtils.parseObject(data, WorkerHeartBeat.class));  //  ???
> > > workerNodeInfo.put(workerAddress, info);  // put "worker" key into the workerNodeInfo
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > The effect is that the `addr` of `getHostWeight` will be the `worker` and `heartBeat == null`.
> > > so
> > > ```java
> > >         if (heartBeat == null) {
> > >             logger.warn("worker {} in work group {} have not received the heartbeat", addr, workerGroup);
> > >             return Optional.empty();
> > >         }
> > > ```
> > 
> > 
> > So we only need to subscribe the event of REGISTRY_DOLPHINSCHEDULER_WORKERS's child path
> 
> Please help to review my code to see if there are other problems. If there is a better implementation, please reply to me. And I have fixed it in my production environment and it works fine.

Does master path will have this problem? could you please change the master together?

lxorc committed:
Merge branch 'dev' into patch-2

ruanwenjun commented:
Please fix the code style by `mvn spotless:apply`

Gallardot commented on dolphinscheduler-registry/dolphinscheduler-registry-api/src/main/java/org/apache/dolphinscheduler/registry/api/RegistryClient.java:
```diff
@@ -210,7 +210,7 @@ public boolean isMasterPath(String path) {
 
     public boolean isWorkerPath(String path) {
-        return path != null && path.startsWith(RegistryNodeType.WORKER.getRegistryPath());
+        return path != null && path.startsWith(Constants.REGISTRY_DOLPHINSCHEDULER_WORKERS) && !path.equals(Constants.REGISTRY_DOLPHINSCHEDULER_WORKERS);
```
```suggestion
        return path != null && path.startsWith(Constants.REGISTRY_DOLPHINSCHEDULER_WORKERS)
                && !path.equals(Constants.REGISTRY_DOLPHINSCHEDULER_WORKERS);
```


@lxorc Do you have time to work on this PR? Maybe I can help with that. We also encountered this issue, and I have tested this PR in our environment. It is feasible.

lxorc committed:
Update dolphinscheduler-registry/dolphinscheduler-registry-api/src/main/java/org/apache/dolphinscheduler/registry/api/RegistryClient.java

Co-authored-by: Gallardot <gallardot@apache.org>
